src/RequestGroupMan.cc

```diff
@@ -1005,4 +1026,58 @@ void RequestGroupMan::decreaseNumActive()
   --numActive_;
 }
 
+
+int RequestGroupMan::optimizeConcurrentDownloads()
+{
+  // gauge the current speed
+  int currentSpeed=getNetStat().calculateDownloadSpeed();
+
+  // estimate the actual bandwidth consumed including connection overheads, assuming 4KiB of overhead for each connection initialisation
+  int bandwidth=currentSpeed;
+  double speedUndervalueRatio(0.);
+  uint64_t dl(getNetStat().getSessionDownloadLength());
+  if(dl>0) {
+    speedUndervalueRatio=4096.*(numActive_+numStoppedTotal_)/dl;
+    bandwidth*=(1+speedUndervalueRatio);
```

Comment:
you are quite wise in raising that question ;-), it made me do some additional tests and these tests made me conclude that i don't have a good answer!
If you are curious, please read along. otherwise you can ignore since my new proposal would be to remove this part. i found out since then that the concept i was trying to model is actually not necessary to meet the objective it was supposed to contribute to...

the original objective was to take into account the 'overhead'  imposed by each of the file transfers on the speed measured, noting that the speed reached by aria2 in the case of small files is much lower compared to that of large files, even in the case of a high number of parallel downloads, which in turns makes the algo conclude on a too small number of parallel dowloads at the next iteration (the case highlighted by luokar).

the rationale for this calculation was hence to compute the actual bandwidth utilized instead of just measuring the effective download speed. As a matter of fact, the 'speed' computed by aria2 in SpeedCalc is measuring the actual number of 'useful' file bytes received by the client in a second of time after the first 'useful' byte has been received for every file download active, which is not necessarily linear with the actual bandwidth utilized if one is to consider the initial traffic necessary to initialise each transfer (e.g.ftp or http connection initialisation overhead) before the first 'useful' byte is received.
 
The speedUndervalueRatio was here to measure this overhead on the principle that every file transfer, regardless of the size of the file, would impose a fixed traffic on the link of N initialisation bytes before the first useful byte is received and accounted for in SpeedCalc.
Along this principle:
 speedUndervalueRatio = bandwidth/speed-1 
= (bytesWindow_ + N*(number of downloads active in the speed elapsed period))/bytesWindow_ -1 
= N*(number of downloads active in the speed elapsed period)/bytesWindow_

in turn, and on the assumption that the overhead would have limited variability in a session (sizes of files equally distributed along the session time), the overhead could be approximated as:
speedUndervalueRatio ~ N*(total number of downloads since the start of the session)/(total number of bytes received since the start of the session)

after my empirical tests, i had concluded for a value for N in the order of 4 KiB so that the bandwidth computed on 1KiB parallel ftp file transfers could be comparable to that of 1MiB parallel ftp file transfers.

---------------------
Now, back to reality after my additional tests:
- i found out that in case of http, the overhead would need to be much less than the 4KiB which i had benchmarked on ftp transfers. 
- trying to reach a higher number of parallel downloads for small files as compared to the case of large files is not necessarily a good thing since it can tire up the server(s) which have to process too many requests too frequently (the frequency of requests in a scenario with many small files is larger since they are downloaded quickly)
- finally and more importantly, the main unwanted scenario in the case of small files is already taken care of by the Timer of 5 seconds that waits for the parallel downloads to ramp-up the speed while ignoring possible drops of speed while the steady state is being reached. Also, monitoring that the speed keeps up well before decreasing number of parallel downloads is warrant that the steady state is maintained even for small file downloads.

in conclusion, i propose to roll back to using the speed directly which gives equally good results in both ftp and http cases.